### DISCUSSION TITLE
Model vs CompositeModel class

### DISCUSSION THREAD

**Discussion by David-HERS** [lmfit/lmfit-py#863] [Q&A] [answered]
What are the advantages and disadvantages of creating my own model or creating a composite model?
talking about optimization and not the ease of creation that comes from using a composite model. Will asteval's function end up creating the same function that you would define manually? and there would be no differences?.
An answer to further my understanding would be highly appreciated!

**Comment by newville** [comment, marked as answer]
Using a CompositeModel instead of hand-coded Model is really a matter of convenience.   Asteval is used only if you use an ExpressionModel (which should be viewed as a "cute trick", and maybe useable, but not really recommended) and for implementing constraints. 

The resulting route through Python code will be different for a Model function that implements the same functionality as a CompositeModel, but the mathematics of the solution (basically down to the number of function evaluations and steps in the parameter values) should be identical.  

Just to be clear, the entire rationale for `lmfit` existing at all, and the entire rationale for `scipy` existing at all can be described as "convenience".  The actual math can be done with any Turiing-complete system, and if runtime is the only consideration and shaving microseconds or kilobytes of RAM is the priority, you're in the wrong place.  And you're optimizing the wrong resource.  We (Python, numpy, scipy, lmfit) are optimizing user time and memory, not computer time or memory.
